wire-compiler/src/main/java/com/squareup/wire/model/ProtoTypeName.java

```diff
@@ -0,0 +1,123 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.model;
+
+import java.lang.reflect.Field;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.squareup.wire.model.Util.checkNotNull;
+import static com.squareup.wire.model.Util.checkState;
+
+/**
+ * Names a protocol buffer message, enumerated type or a scalar. This class models a fully-qualified
+ * name using the protocol buffer package.
+ */
+public final class ProtoTypeName {
+  public static final ProtoTypeName BOOL = new ProtoTypeName("bool");
+  public static final ProtoTypeName BYTES = new ProtoTypeName("bytes");
+  public static final ProtoTypeName DOUBLE = new ProtoTypeName("double");
+  public static final ProtoTypeName FLOAT = new ProtoTypeName("float");
+  public static final ProtoTypeName FIXED32 = new ProtoTypeName("fixed32");
+  public static final ProtoTypeName FIXED64 = new ProtoTypeName("fixed64");
+  public static final ProtoTypeName INT32 = new ProtoTypeName("int32");
+  public static final ProtoTypeName INT64 = new ProtoTypeName("int64");
+  public static final ProtoTypeName SFIXED32 = new ProtoTypeName("sfixed32");
+  public static final ProtoTypeName SFIXED64 = new ProtoTypeName("sfixed64");
+  public static final ProtoTypeName SINT32 = new ProtoTypeName("sint32");
+  public static final ProtoTypeName SINT64 = new ProtoTypeName("sint64");
+  public static final ProtoTypeName STRING = new ProtoTypeName("string");
+  public static final ProtoTypeName UINT32 = new ProtoTypeName("uint32");
+  public static final ProtoTypeName UINT64 = new ProtoTypeName("uint64");
+
+  private static final Map<String, ProtoTypeName> SCALAR_TYPES;
+  static {
+    Map<String, ProtoTypeName> scalarTypes = new LinkedHashMap<String, ProtoTypeName>();
+    try {
+      for (Field field : ProtoTypeName.class.getDeclaredFields()) {
+        if (field.getType() == ProtoTypeName.class) {
+          ProtoTypeName scalar = (ProtoTypeName) field.get(null);
+          scalarTypes.put(scalar.names.get(0), scalar);
+        }
+      }
+      SCALAR_TYPES = Collections.unmodifiableMap(scalarTypes);
+    } catch (IllegalAccessException e) {
+      throw new AssertionError();
+    }
+  }
+
+  private final String protoPackage;
+  private final List<String> names;
+  private final boolean isScalar;
+
+  private ProtoTypeName(String scalarName) {
+    this(null, Collections.singletonList(scalarName), true);
+  }
+
+  private ProtoTypeName(String protoPackage, List<String> names, boolean isScalar) {
+    this.protoPackage = protoPackage;
+    this.names = names;
+    this.isScalar = isScalar;
+  }
+
+  public String simpleName() {
+    return names.get(names.size() - 1);
```

Comment:
Rather than plaicing the simple name by convention in `names.get(names.size() - 1)`, could it be a separate instance variable for clarity?